### PR TITLE
Deprecate device-specific GradScaler autocast API

### PR DESCRIPTION
**Pull request by guangyey**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126531
* __->__ #126527

# Motivation

## for `torch.amp.GradScaler`,
- `torch.cpu.amp.GradScaler(args...)` is completely equivalent to `torch. amp.GradScaler("cpu", args...)`.
- `torch.cuda.amp.GradScaler(args...)` is completely equivalent to `torch.amp.GradScaler("cuda", args...)`.

So, we intend to depreate them and **strongly recommend** developer to use `torch.amp.GradScaler`.

## for `custom_fwd` and `custom_bwd`,
this is a good solution to make the custom function run with or without effect even in an autocast-enabled region and can be shared by other backends, like CPU and XPU.
So we generalize it to be device-agnostic and put them int `torch/amp/autocast_mode.py` and re-expose to `torch.amp.custom_fwd` and `torch.amp.custom_bwd`. Meanwhile, we deprecate `torch.cuda.amp.custom_fwd` and `torch.cuda.amp.custom_bwd`.

# Additional Context
Add UT to cover the deprecated warning.
No need for more UTs to cover the functionality of `torch.amp.custom_f/bwd`, the existing UTs that previously covered the functionality of `torch.cuda.amp.custom_f/bwd` can cover them.
To facilitate the review, we separate these code changes to two PRs. The first PR cover `torch.amp.GradScaler`. The follow-up covers `custom_fwd` and `custom_bwd`.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @ezyang @gchanan @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mcarilli @ptrblck @leslie-fang-intel @voznesenskym @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng

### CODE REVIEW

**Review by jgong5 — APPROVED** [review]

**Review by gujinghui — APPROVED** [review]

**Review by janeyx99 — CHANGES_REQUESTED** [review]
This deprecation is desirable, but we should remove all mentions of torch.cuda.amp.GradScaler and torch.cpu.amp.GradScaler in the codebase (e.g., in the tests) and replace with the new usage.

This is one way to ensure that the recommended version of GradScaler will remain sufficiently tested too.

File: test/test_torch.py [hunk 1]
```diff
@@ -6152,7 +6152,8 @@ def run(model0, model1, optimizer0, optimizer1, try_scaling_api):
     @onlyNativeDeviceTypes
     def test_grad_scaler_pass_itself(self, device):
         device = torch.device(device)
-        GradScaler = torch.cuda.amp.GradScaler if "cuda" == device.type else torch.cpu.amp.GradScaler
+        device_type = "cuda" if "cuda" == device.type else "cpu"
```
Comment (janeyx99) [inline test/test_torch.py:6155]:
```suggestion
```
is this line still needed?

File: test/test_torch.py [hunk 2]
```diff
@@ -6152,7 +6152,8 @@ def run(model0, model1, optimizer0, optimizer1, try_scaling_api):
...
+        GradScaler = partial(torch.amp.GradScaler, device=device_type)
```
Comment (janeyx99) [inline test/test_torch.py:6156]:
```suggestion
        GradScaler = partial(torch.amp.GradScaler, device=device)
```
should work?

**Review by janeyx99 — COMMENTED** [review]
Please also update the docs https://github.com/pytorch/pytorch/blob/86ad101370075cfa392c953b17e1a8ab3e66881f/docs/source/notes/amp_examples.rst#L112

File: test/test_torch.py (hunk 2 above)
Comment (guangyey) [reply test/test_torch.py]: here defines `onlyNativeDeviceTypes`
https://github.com/pytorch/pytorch/blob/86ad101370075cfa392c953b17e1a8ab3e66881f/torch/testing/_internal/common_utils.py#L230
For `meta` and `PU1` we should test it on `cpu`.

**Comment by guangyey** [comment]
> Please also update the docs
> 
> https://github.com/pytorch/pytorch/blob/86ad101370075cfa392c953b17e1a8ab3e66881f/docs/source/notes/amp_examples.rst#L112

I update deprecation warning [here](https://github.com/pytorch/pytorch/pull/126527/files#diff-b17e8e6fc2ce9966e0a0b1a7d322a2a5938deb671b2e6f1c471ae1a5937a4dd4R11), I think it will be exposed here https://pytorch.org/docs/stable/amp.html#gradient-scaling feature. May I know if I capture your point?

File: test/test_torch.py (hunk 2 above)
Comment (guangyey) [reply test/test_torch.py]: following your suggestion, update the code.

**Comment by janeyx99** [comment]
@guangyey I mean that this page should also get updated to direct people to the recommended API
image

**Comment by guangyey** [comment]
> @guangyey I mean that this page should also get updated to direct people to the recommended API image

I updated the doc. Could you help review this PR again?

**Comment by guangyey** [comment]
@janeyx99  could you help review this PR again? Thanks very much~

**Review by EikanWang — APPROVED** [review]

File: docs/source/amp.rst [hunk 3]
```diff
@@ -25,12 +25,9 @@ However, :class:`torch.autocast` and :class:`torch.GradScaler` are modular, and
...
 datatype of ``torch.bfloat16`` only uses :class:`torch.autocast`.
 
-For CUDA and CPU, APIs are also provided separately:
-
-* ``torch.autocast("cuda", args...)`` is equivalent to ``torch.cuda.amp.autocast(args...)``.
-* ``torch.autocast("cpu", args...)`` is equivalent to ``torch.cpu.amp.autocast(args...)``. For CPU, only lower precision floating point datatype of ``torch.bfloat16`` is supported for now.
-* ``torch.GradScaler("cuda", args...)`` is equivalent to ``torch.cuda.amp.GradScaler(args...)``.
-* ``torch.GradScaler("cpu", args...)`` is equivalent to ``torch.cpu.amp.GradScaler(args...)``.
+.. warning::
+    ``torch.cuda.amp.autocast(args...)`` and ``torch.cpu.amp.autocast(args...)`` will be deprecated. Please use ``torch.autocast("cuda", args...)`` or ``torch.autocast("cpu", args...)`` instead.
```
Comment (janeyx99) [inline docs/source/amp.rst:29]: Include this fact here still?
For CPU, only lower precision floating point datatype of ``torch.bfloat16`` is supported for now.

File: docs/source/amp.rst [hunk 4]
```diff
@@ -25,12 +25,9 @@ However, :class:`torch.autocast` and :class:`torch.GradScaler` are modular, and
 As shown in the CPU example section of :class:`torch.autocast`, "automatic mixed precision training/inference" on CPU with
 datatype of ``torch.bfloat16`` only uses :class:`torch.autocast`.
 
-For CUDA and CPU, APIs are also provided separately:
```
Comment (janeyx99) [inline docs/source/amp.rst:28]: Replace mention of torch.cpu.amp.GradScaler and cuda.amp.GradScaler with just amp.Scaler in line 22

**Review by janeyx99 — APPROVED** [review]
Some last nits, thanks!

File: docs/source/amp.rst (hunk 4 above)
Comment (guangyey) [reply docs/source/amp.rst:28]: Updated.

File: docs/source/amp.rst (hunk 3 above)
Comment (guangyey) [reply docs/source/amp.rst:29]: I remember that `torch.float16` is already supported on CPU now, right @leslie-fang-intel

Comment (leslie-fang-intel) [reply docs/source/amp.rst:29]: I think CPU Autocast also support `torch.float16` now

**Comment by guangyey** [comment]
> Some last nits, thanks!

Thanks for your approval. Have a nice day~

**Comment by kit1980** [comment]
This is bc-breaking BTW (and actually broke some Meta-internal stuff).
I've added bc-breaking label.

Also probably this should be added to TorchFix https://github.com/pytorch-labs/torchfix

**Comment by kit1980** [comment]
I removed bc-breaking as it turned out it was another PR that broke BC.
